### PR TITLE
feat: add string/base/distances/levenshtein

### LINKED ISSUE #835: [RFC]: Levenshtein distance between two strings
### Description

This RFC proposes adding a function to calculate the Levenshtein (edit) distance between two strings.

Package: `@stdlib/string/base/distances/levenshtein`.

The function should have the following signature `(a: string, b: string): number`.

The Levenstein (edit) distance is a popular metric for quantifying the difference between two strings. It is used in a variety of applications, including spell checking, natural language processing, and text mining.

See:
 - [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Infinage**
Resolves #835.

## Description

> What is the purpose of this pull request?

This pull request:

-   Implements the RFC proposal for implementing function to calculate the Levenshtein (edit) distance between two strings.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #835 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/string/base/distances/README.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+# Distances
+
+> Implementations of various string similarity metrics.
+
+<section class="usage">
+
+## Usage
+
+```javascript
+var string = require( '@stdlib/string/base/distances' );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/README.md]: Let's name the variable distances here and below in the namespace `README.md`.

File: lib/node_modules/@stdlib/string/base/distances/docs/types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+/**
+* Implementations of various string similarity metrics.
+*/
+interface Namespace {
+	/**
+	* Calculates the Levenshtein (edit) distance between two strings.
+	*
+	* @param str1 - first string value
+    * @param str2 - second string value
+    * @returns levenshtein distance
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/docs/types/index.d.ts]:
```suggestion
    * @returns Levenshtein distance
```

File: lib/node_modules/@stdlib/string/base/distances/examples/index.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+var ns = require( './../lib' );
+
+// Calculates the levenshtein (edit) distance between two input strings...
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/examples/index.js]:
```suggestion
// Calculates the Levenshtein (edit) distance between two input strings...
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/docs/types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+*/
+
+// TypeScript Version: 2.0
+
+/**
+* Calculates the Levenshtein (edit) distance between two strings.
+*
+* @param str1 - first string value
+* @param str2 - second string value
+* @returns levenshtein distance
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/docs/types/index.d.ts]:
```suggestion
* @returns Levenshtein distance
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+
+// MAIN //
+
+/**
+* Calculates the Levenshtein (edit) distance between two strings.
+*
+* @param {string} s1 - first string value
+* @param {string} s2 - second string value
+* @throws {TypeError} both arguments provided must be a string
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js]:
```suggestion
* @throws {TypeError} first argument must be a string
* @throws {TypeError} second argument must be a string
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+* @returns {number} levenshtein distance
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js]:
```suggestion
* @returns {number} Levenshtein distance
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	if ( !isString( s2 ) ) {
+		throw new TypeError( format( 'invalid argument. Must provide a string. Value: `%s`.', s2 ) );
+	}
+
+	n = s1.length;
+	m = s2.length;
+	row = [];
+
+	// If either of the string is empty, insert the characters
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js]:
```suggestion
	// If either of the strings is empty, insert the characters:
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/test/test.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.strictEqual( typeof levenshteinDistance, 'function', 'main export is a function' );
+	t.end();
+});
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/test/test.js:34]: A few unit tests are missing for full code coverage. Should add tests for invalid inputs (see e.g. `@stdlib/string/substring-after` for a function with two string parameters).

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Thanks for the PR! Only a few minor things need to be addressed.

**Comment by Infinage** [comment]
Thanks for the quick review! Had a minor query - post making these changes, should I do a squash and force push these changes or do you want me to retain the commit history?

**Comment by Infinage** [comment]
@Planeshifter - I have force pushed with the requested changes. Could you please approve it if everything looks good? Thanks!

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/docs/types/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/docs/types/index.d.ts]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/docs/types/test.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/docs/types/test.ts]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/examples/index.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/examples/index.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/test/test.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/test/test.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/lib/index.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/lib/index.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/index.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/index.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/levenshtein/lib/main.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/examples/index.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/examples/index.js]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/docs/types/test.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/docs/types/test.ts]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/docs/types/index.d.ts [hunk 18]
```diff
@@ -0,0 +1,61 @@
+/*
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/docs/types/index.d.ts]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/string/base/distances/README.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
+<!--
+
+@license Apache-2.0
+
+Copyright (c) 2022 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/string/base/distances/README.md]:
```suggestion
* Copyright (c) 2023 The Stdlib Authors.
```

**Review by Planeshifter — CHANGES_REQUESTED** [review]

**Review by Planeshifter — APPROVED** [review]

**Review by Planeshifter — APPROVED** [review]

**Comment by Planeshifter** [comment]
Package has landed! Thanks again.

**Comment by Infinage** [comment]
Truly elated I could help! Thank you!

**Pull request merged**